### PR TITLE
Chapter 2

### PR DESCRIPTION
**Pull request by thanakritju**
### Solutions for Chapter {NUMBER}

cc @vrom911 @chshersh

### CODE REVIEW

**Comment by chshersh** [comment]
Hi @thanakritju! I see that you've merged this PR, but you've mentioned our names. Would you still like to have our review? :slightly_smiling_face:

**Comment by thanakritju** [comment]
@chshersh please if you could provide me some :) , not sure if is there is a way to contribute back to you or even the haskell community, I really enjoy your work. Thanks again :)

File: src/Chapter2.hs [hunk 1]
```diff
@@ -336,7 +336,9 @@ from it!
 ghci> :l src/Chapter2.hs
 -}
 subList :: Int -> Int -> [a] -> [a]
-subList = error "subList: Not implemented!"
+subList startIndex endIndex array
+    | startIndex < 0 || endIndex < 0 = []
```
Comment (chshersh) [inline src/Chapter2.hs:340]: Your implementation is correct when the right bound is lower than the left. But I think you can implement it more efficiently by adding a separate case to your check :slightly_smiling_face:

File: src/Chapter2.hs [hunk 2]
```diff
@@ -606,7 +614,8 @@ Implement a function that duplicates each element of the list
 
 -}
 duplicate :: [a] -> [a]
-duplicate = error "duplicate: Not implemented!"
+duplicate [] = []
+duplicate (x:xs) = [x, x] ++ duplicate xs
```
Comment (chshersh) [inline src/Chapter2.hs:618]: It is more efficient to construct lists using the cons operator directly without introducing intermediate values.
```suggestion
duplicate (x:xs) = x : x : duplicate xs
```

File: src/Chapter2.hs [hunk 3]
```diff
@@ -741,7 +754,8 @@ the list with only those lists that contain a passed element.
 
 🕯 HINT: Use the 'elem' function to check whether an element belongs to a list
 -}
-contains = error "contains: Not implemented!"
+contains :: Int -> [[Int]] -> [[Int]]
+contains n l = filter (\x -> elem n x) l
```
Comment (chshersh) [inline src/Chapter2.hs:758]: Now, after you've mastered eta-reduction, you can apply such a technique to this function as well  🙂 
But your implementation is already great 💯

Here you have two opportunities for eta-reduction: lambda and the last argument to `filter` :recycle:

File: src/Chapter2.hs [hunk 4]
```diff
@@ -858,7 +877,10 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [Int] -> [Int]
+rewind [] = []
+rewind [x] = [x]
+rewind (x:xs) = (rewind xs) ++ [x]
```
Comment (chshersh) [inline src/Chapter2.hs:883]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

File: src/Chapter2.hs [hunk 5]
```diff
@@ -842,7 +858,10 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [Int] -> [Int]
+rotate n l
+    | n < 0 = []
+    | otherwise = take (length l) (drop n (cycle l))
```
Comment (chshersh) [inline src/Chapter2.hs:864]: That's a correct solution! Though, it does a little bit more than work when the given number is big and the list is small. You will cycle through the list a lot. Maybe you can think of a way to optimize this :slightly_smiling_face:

**Review by chshersh — COMMENTED** [review]
Nice work on solving Chapter Two challenges!
And well done with the Advanced Task :medal_sports: 

I left my review and some comments. Hope you find them helpful :slightly_smiling_face:

You can contribute back in many ways! :infinity: You can [support us on Ko-Fi](https://ko-fi.com/kowainik) or sponsor us on GitHub:

* https://github.com/sponsors/vrom911
* https://github.com/sponsors/chshersh

If you're interested, you can also contribute to various Haskell open-source projects, this is always appreciated :pray:

**Pull request merged**